global_controller/global_controller_magma.py

```diff
@@ -6,98 +6,125 @@
 from gemstone.generator.generator import Generator
 from . import global_controller_genesis2
 from global_buffer.mmio_type import MMIOType
+from global_controller.axi4_type import AXI4SlaveType
 
 
 class GlobalController(Generator):
-    def __init__(self, addr_width, data_width, soc_addr):
+    def __init__(self, addr_width, data_width):
         super().__init__()
 
         self.addr_width = addr_width
         self.data_width = data_width
-        self.soc_addr = soc_addr
         self.config_type = ConfigurationType(self.addr_width, self.data_width)
-        self.top_config_type = ConfigurationType(self.soc_addr,
-                                                 self.data_width)
         self.glb_config_type = ConfigurationType(self.addr_width,
                                                  self.data_width)
 
         self.add_ports(
-            jtag=JTAGType,
-            config=magma.Out(self.config_type),
-            read_data_in=magma.In(magma.Bits[self.data_width]),
-            top_config=magma.Out(self.top_config_type),
-            top_read_data_in=magma.In(magma.Bits[self.data_width]),
-            glb_config=magma.Out(self.glb_config_type),
-            glb_read_data_in=magma.In(magma.Bits[self.data_width]),
             clk_in=magma.In(magma.Clock),
             reset_in=magma.In(magma.AsyncReset),
+
             clk_out=magma.Out(magma.Clock),
             reset_out=magma.Out(magma.AsyncReset),
+            stall=magma.Out(magma.Bits[4]),
+            glb_stall=magma.Out(magma.Bit),
+
             cgra_start_pulse=magma.Out(magma.Bit),
             cgra_done_pulse=magma.In(magma.Bit),
+            cgra_soft_reset=magma.Out(magma.Bit),
```

Comment:
As far as I know, the CGRA only has 1 reset... is this another global signal we need to add?

Reply:
That is `soft_reset`, which is not resetting registers but resetting some counters in cgra.
You can think of it as 1bit input, instead of `reset`